### PR TITLE
get users and assorted Okta changes

### PR DESCRIPTION
**Pull request by jeremyzitomer-usds**
## Related Issue or Background Info

Added mutation to get all users with role information

## Changes Proposed

- Added mutation
- Renamed/refactored some internal classes; notably, OrganizationRoles -> AuthorityBasedOrganizationRoles, and CurrentOrganizationRoles -> OrganizationRoles
- Updated User class
- Added new permission "MANAGE_USER_LIST"
- removed Okta integration from API tests
- added some demo/test Okta mocking

## Rationale for the OrganizationRoles renames
I realized that the way that the old "OrganizationRoles" class was really being used, was exclusively when grabbing an authority/claim from the JWT or Okta (what those external sources provide is always an org's external ID and the associated user roles). Hence the rationale to call these roles "authority-based". I recognize there's a bit of grey between authority and claims, if there's a better word for this, I'm all ears.
Also, the old "CurrentOrganizationRoles" class was identical to the object that was needed in order to return organization roles for users of other organizations (especially in the create/updateUser flows), and it became too limiting to require that such a class be used only to denote org roles for the caller's current org.

### CODE REVIEW

File: backend/src/main/java/gov/cdc/usds/simplereport/api/apiuser/ApiUserResolver.java [hunk 1]
```diff
@@ -22,15 +28,35 @@
...
-		Optional<CurrentOrganizationRoles> currentOrgRoles = _organizationService.getCurrentOrganizationRoles();
-		Boolean isAdmin = _userService.isAdminUser(currentUser);
+		Optional<OrganizationRoles> currentOrgRoles = _organizationService.getCurrentOrganizationRoles();
+		Boolean isAdmin = _userService.isAdmin(currentUser);
         return new User(currentUser, currentOrgRoles, isAdmin);
 	}
+
+	public List<User> getUsers() {
+		Organization org = _organizationService.getCurrentOrganization();
+		List<ApiUser> users = _organizationService.getUsersInCurrentOrg(OrganizationRole.USER);
```
Comment (sharmaneil) [inline backend/src/main/java/gov/cdc/usds/simplereport/api/apiuser/ApiUserResolver.java:45]: Would this only get users with the role USER, or everyone? From a quick scan, the `OrganizationRole.USER` parameter is used to query the okta group -- are only users with role USER in that group?

I ask because i assume the graphql query for `users` would return all humans, not just USER humans.

Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/apiuser/ApiUserResolver.java:45]: it gets everyone. everyone in an organization is in the USER Okta group for that org. what the lines after that do is assign the specific roles of all users based on the additional Okta groups (ADMIN, ENTRY_ONLY) they are in for that organization.

File: backend/build_and_push.sh [hunk 2]
```diff
@@ -13,5 +13,5 @@ echo "Tagged $ACR_TAG"
 
 # Log in to ACR
 # Uncomment for user login
-# az acr login --name simplereportacr
+az acr login --name simplereportacr
```
Comment (timothybest-usds) [inline backend/build_and_push.sh:16]: Is this change intentional?

Comment (jeremyzitomer-usds) [reply backend/build_and_push.sh]: nope! thanks for catching :)

**Comment by benwarfield-usds** [comment]
Thanks for the extra information on the renames—I agree that the swap is logical, though more than a little confusing on first read. 😁 

Based on your (correct!) description of what the original OrganizationRoles object does, I would call it `OrganizationRoleClaims` or `OrganizationClaims` depending on whether or not we think the same object will eventually house the facility-access information (I think probably yes, but it'll require a refactor if so).

I don't think using Authority in the name is helpful: it's only Authority-based as a matter of convenience based on the way Okta sets up their JWT integration, and once we move beyond the current (comparatively) simple role-management approach, we will have to move beyond (or possibly entirely abandon) the `getAuthorities` approach to querying the Authorization object.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/OktaServiceConfig.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    @Bean
+    @Profile(BeanProfiles.NO_OKTA_MGMT)
+    public OktaService getDummyOktaService(InitialSetupProperties setupProps, DemoUserConfiguration demoUsers) {
+        Map<String,OrganizationRoleClaims> usernameRolesMap = new HashMap<>();
+        for (DemoAlternateUser altUser : demoUsers.getAlternateUsers()) {
+            IdentityAttributes user = altUser.getIdentity();
+            String username = user.getUsername();
+            OrganizationRoleClaims orgRoles = 
+                    new OrganizationRoleClaims(setupProps.getOrganization().getExternalId(),
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/OktaServiceConfig.java:40]: Couldn't this just be `demoUsers.getDefaultUser().getAuthorization().getOrganizationExternalId()`?

**Review by benwarfield-usds — COMMENTED** [review]
I had some structural questions but we resolved those offline, and I think I've deleted all the comments related to them.

I had a lot of trouble getting a consecutive time block to do the whole thing and get my head around it, so my apologies for the slowness that caused—it's a lot of individual changes to navigate around, even if the straight line-count isn't as big as some.

I made a bunch of fairly small-scale comments that I am mostly just assuming you can figure out which of seem significant and which not—the major points I wanted to raise and which may not be obvious from what I said inline are:

* I'm confused about the way the demo-user configuration is being used to create fake-Okta, and how it's entangled with the OrganizationInitializationService: I think that's worth teasing apart in order to get to where we will need to get to on having a good local dev/demo setup. Obviously trying to get to a fully-fleshed-out multi-role demo is not in scope for this PR, but I think a little bit of work to keep those things from getting tangled up (especially since I just recently untangled them) is worthwhile—the suggestions I made around DemoUserConfiguration and EmptyOktaServiceImpl were aimed at that, but I'd be happy to discuss offline as well.
* I'm concerned that the tests that set up the user to be a superuser _and_ an org admin indicate that our permission model for those endpoints is broken, and I would like us to fix that ASAP. I'm not sure that's a blocker either, for this PR, but we need to prioritize that sometime sooner than "after we have nothing else to do."

File: backend/src/test/java/gov/cdc/usds/simplereport/service/BaseServiceTest.java [hunk 4]
```diff
@@ -108,8 +107,7 @@ protected void clearOktaUsers() {
     protected void clearOktaGroups() {
         for (Group g : _oktaClient.listGroups()) {
             String groupName = g.getProfile().getName();
-            if (g.getType() == GroupType.OKTA_GROUP &&
-                        groupName.startsWith(_authorizationProperties.getRolePrefix())) {
+            if (groupName.startsWith(_authorizationProperties.getRolePrefix())) {
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/service/BaseServiceTest.java:110]: not sure how i removed it, my bad!

[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/service/BaseServiceTest.java:110]: (not that the entire code block will stick around much longer anyway)

File: backend/src/main/java/gov/cdc/usds/simplereport/api/model/User.java [hunk 5]
```diff
@@ -1,41 +1,35 @@
...
+
+    public User(ApiUser user, Optional<OrganizationRoles> orgwrapper, boolean isAdmin) {
         this.wrapped = user;
-        this.org = orgwrapper.map(CurrentOrganizationRoles::getOrganization);
+        this.org = orgwrapper.map(OrganizationRoles::getOrganization);
         this.permissions = new ArrayList<>();
-        this.isAdmin = isAdmin;
-        Optional<OrganizationRole> effectiveRole = orgwrapper.flatMap(CurrentOrganizationRoles::getEffectiveRole);
+        this.roles = new ArrayList<>();
+        orgwrapper.map(OrganizationRoles::getGrantedRoles).map(roles::addAll);
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/model/User.java:28]: i want to return the full role list. while it is currently true that the permissions for our roles are concentric circles around each other, that may not always be true, so returning multiple may be valuable and i'd rather not shoehorn us.

File: backend/src/main/java/gov/cdc/usds/simplereport/api/apiuser/ApiUserResolver.java [hunk 6]
```diff
@@ -22,15 +28,35 @@
...
 	public User getWhoami() {
 		ApiUser currentUser = _userService.getCurrentUser();
-		Optional<CurrentOrganizationRoles> currentOrgRoles = _organizationService.getCurrentOrganizationRoles();
-		Boolean isAdmin = _userService.isAdminUser(currentUser);
+		Optional<OrganizationRoles> currentOrgRoles = _organizationService.getCurrentOrganizationRoles();
+		Boolean isAdmin = _userService.isAdmin(currentUser);
         return new User(currentUser, currentOrgRoles, isAdmin);
 	}
+
+	public List<User> getUsers() {
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/apiuser/ApiUserResolver.java:43]: yeah i couldn't find a good way to push this down the stack without pushing the User api model object down the stack as well, which smelled bad. *shrug*

File: backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/UserPermission.java [hunk 7]
```diff
@@ -10,6 +10,7 @@
     EDIT_PATIENT,
     EDIT_FACILITY,
     EDIT_ORGANIZATION,
+    MANAGE_USER_LIST,
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/UserPermission.java:13]: is there a reason to make it MANAGE_USERS over MANAGE_USER_LIST ? just for extra differentiation?

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java [hunk 8]
```diff
@@ -105,6 +112,19 @@ public void assertFacilityNameAvailable(String testingFacilityName) {
         return _facilityRepo.findByOrganizationOrderByFacilityName(org);
     }
 
+    @AuthorizationConfiguration.RequirePermissionManageUserList
+    public List<ApiUser> getUsersInCurrentOrg(OrganizationRole role) {
+        String organizationExternalId = getCurrentOrganization().getExternalId();
+        List<String> usernames = _oktaService.getAllUsernamesForOrganization(organizationExternalId, role);
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:118]: will call the below method, will leave the second refactor point for my next PR since I've started that anyway

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java [hunk 9]
```diff
@@ -86,12 +91,14 @@ public Organization getOrganization(String externalId) {
...
     }
 
-    public Optional<Organization> getOrganizationForUser(ApiUser apiUser) {
-        String orgExternalId = _oktaService.getOrganizationExternalIdForUser(apiUser.getLoginEmail());
-        if (orgExternalId == null) {
-            return Optional.empty();
-        }
-        return Optional.ofNullable(getOrganization(orgExternalId));
+    @AuthorizationConfiguration.RequirePermissionManageUserList
+    public Optional<OrganizationRoles> getOrganizationRolesForUser(ApiUser apiUser) {
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:95]: agreed, which is why i've already started them moving them over in my next PR 😄 will address this then.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java [hunk 10]
```diff
@@ -37,25 +39,28 @@
     private OrganizationRepository _repo;
     private FacilityRepository _facilityRepo;
     private ProviderRepository _providerRepo;
+    private ApiUserRepository _userRepo;
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:42]: ditto re next PR.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java [hunk 11]
```diff
@@ -123,25 +147,37 @@ public void deleteOrganization(String externalId) {
     }
 
     // returns the external ID of the organization the specified user belongs to
-    public String getOrganizationExternalIdForUser(String username) {
+    public Optional<OrganizationRoleClaims> getOrganizationRolesForUser(String username) {
+
         UserList users = _client.listUsers(username, null, null, null, null);
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java:152]: Okta 4 evr <3

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java [hunk 12]
```diff
@@ -123,25 +147,37 @@ public void deleteOrganization(String externalId) {
...
+                    orgExternalIds.add(getOrganizationExternalIdFromGroupName(groupName, role));
+                    roles.add(role);
+                    break;
+                }
             }
         }
 
-        throw new IllegalGraphqlArgumentException("User is not in any Okta user groups");
+        if (orgExternalIds.size() != 1) {
+            LOG.warn("User is in {} Okta organizations, not 1", orgExternalIds.size());
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java:175]: eek, there's no notion of archiving groups in Okta. Either they live on (which feels not good, because then do the old user-group relationships remain? and then does the OktaService always return all groups--existing and archived--that the user belongs to?) or we get rid of them. I'd prefer the latter, which means this code stays as-is. let me know if you feel strongly.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java [hunk 13]
```diff
@@ -84,7 +93,22 @@ public void createUser(IdentityAttributes userIdentity, String organizationExter
                 .buildAndCreate(_client);
     }
 
+    public List<String> getAllUsernamesForOrganization(String externalId, OrganizationRole role) {
+        String groupName = generateGroupName(externalId, role);
+
+        GroupList groups = _client.listGroups(groupName, null, null);
+        if (!groups.iterator().hasNext()) {
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java:100]: i know, i confessed my sins after writing this, but they're not really mine. alas.

File: backend/src/main/java/gov/cdc/usds/simplereport/api/model/User.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/model/User.java:28]: Makes sense. If we end up with overlapping circles we'll need to make sure we are careful about the assumption that USER is still the God-group.

File: backend/src/main/java/gov/cdc/usds/simplereport/api/apiuser/ApiUserResolver.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/apiuser/ApiUserResolver.java:43]: It may become more clear what's possible after other refactors are done and/or as other new code lands. No sense trying to move things around just for the sake of moving them, though.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/UserPermission.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/UserPermission.java:13]: Well, it seems like this role will probably be empowered to edit the settings of individual users, not just the list of them. I'm not actually sure what managing the list means—managing roles, but not being able to edit users?

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java [hunk 14]
```diff
@@ -84,7 +93,22 @@ public void createUser(IdentityAttributes userIdentity, String organizationExter
                 .buildAndCreate(_client);
     }
 
+    public List<String> getAllUsernamesForOrganization(String externalId, OrganizationRole role) {
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java:96]: i don't love that this change passes a db model object into the OktaService class, but i think your upsides outweigh the downsides.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceImpl.java:175]: I feel strongly that we shouldn't have two conflicting definitions of "valid" in different places in the app—I think there are reasonable justifications for the approach that I took originally (well, obviously I do), though I also see your point about cleaning up quickly—worst case this won't be relevant for a significant while, so let's talk about it next week.

File: backend/src/main/java/gov/cdc/usds/simplereport/api/model/User.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/model/User.java:28]: also, there may be some confusion here about orgwrapper being an Optional, not a List, but either way I'll clean it up to make Sonar less mad.

File: backend/src/test/java/gov/cdc/usds/simplereport/api/BaseApiTest.java [hunk 15]
```diff
@@ -182,6 +129,23 @@ protected ObjectNode runQuery(String queryFileName) {
         }
     }
 
+    /**
+     * Run the query in the given resource file, check if the response has the
+     * expected error (either none or a single specific error message), and return
+     * the {@code data} section of the response if the error was as expected.
+     */
+    protected ObjectNode runQuery(String queryFileName, String expectedError) {
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/BaseApiTest.java:137]: is it okay to pass in `variables` as null?

[the remark this answers is not in the document]
Comment (benwarfield-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/BaseApiTest.java:137]: I don't know either, actually. My guess is that it's fine but I'd want to check, clearly.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceEmptyImpl.java [hunk 16]
```diff
@@ -1,36 +1,54 @@
...
+    Map<String,OrganizationRoleClaims> usernameRolesMap;
+    Map<String,Map<OrganizationRole,List<String>>> orgRoleUsernamesMap;
+
+    public OktaServiceEmptyImpl(Map<String,OrganizationRoleClaims> usernameRolesMap) {
+        this.usernameRolesMap = usernameRolesMap;
+        this.orgRoleUsernamesMap = new HashMap<>();
+        for (Map.Entry<String,OrganizationRoleClaims> entry : usernameRolesMap.entrySet()) {
+            Map<OrganizationRole,List<String>> roleUsernamesMap = 
+                    orgRoleUsernamesMap.getOrDefault(entry.getValue().getOrganizationExternalId(),
+                                                     new HashMap<>());
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceEmptyImpl.java:29]: i am going to gently decline for a) readability and b) because i think this code block doesn't address the case of adding usernames to `roleUsernamesMap` when `roleUsernamesMap` is already non-null; and after we add that back in, what i have now feels a touch more straightforward. considering this is just a demo class, i'm going to leave as is. (though I'll pull out orgId as its own variable, to make things a touch more readable.)

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationInitializingService.java [hunk 17]
```diff
@@ -72,6 +80,14 @@ public void initAll() {
 		Facility defaultFacility = _props.getFacility().makeRealFacility(realOrg, savedProvider, defaultDeviceType, configured);
 		LOG.info("Creating facility {} with {} devices configured", defaultFacility.getFacilityName(), configured.size());
 		_facilityRepo.save(defaultFacility);
+
+		// Abusing the class name "OrganizationInitializingService" a little, but the users are in the org.
+		List<IdentityAttributes> users = _demoUserConfiguration.getAlternateUsers().stream()
+				.map(DemoUserConfiguration.DemoAlternateUser::getIdentity).collect(Collectors.toList());
+		for (IdentityAttributes user : users) {
+			_apiUserRepo.save(new ApiUser(user.getUsername(), user));
+			initOktaUser(user, emptyOrg.getExternalId());
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationInitializingService.java:89]: so this is what Dev is -- its profiles include create-sample-data but NOT no-okta-mgmt. We can decide to change that down the line if we'd like.

File: backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java [hunk 18]
```diff
@@ -83,69 +85,215 @@ void whoami_nobody_okResponses() {
...
-            .thenReturn(_initService.getDefaultOrganizationId());
-
-        ObjectNode addVariables = JsonNodeFactory.instance.objectNode()
-            .put("firstName", "Rhonda")
-            .put("middleName", "Janet")
-            .put("lastName", "Jones")
-            .put("suffix", "III")
-            .put("email", USERNAMES.get(0))
-            .put("organizationExternalId", _initService.getDefaultOrganizationId());
+        useOrgAdmin();
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:114]: this (and the other tests where both superuser and org admin are invoked) is an unintentional byproduct of the various security restrictions on the methods involved in creating/updating users, but it's something i'm addressing remedying in my next PR so i'm going to leave it for then

File: backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java [hunk 19]
```diff
@@ -83,69 +85,215 @@ void whoami_nobody_okResponses() {
...
+            assertEquals(userRetrieved.get("organization").get("externalId").asText(),
+                         userAdded.get("organizationExternalId").asText());
+            assertEquals(OrganizationRole.USER.getGrantedPermissions(), 
+                         extractPermissionsFromUser(userRetrieved));
+        }
+    }
+
+    @Test
+    void getUsers_orgUser_failure() {
+        useSuperUser();
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:214]: ditto

File: backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java [hunk 20]
```diff
@@ -83,69 +85,215 @@ void whoami_nobody_okResponses() {
...
+        useOrgAdmin();
+
+        when(_oktaService.getOrganizationRolesForUser(USERNAMES.get(0)))
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:217]: this will be fixable once i resolve the superuser/orgadmin dual dependency in the next PR.

[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:217]: the perils of copypasta

File: backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java [hunk 21]
```diff
@@ -83,69 +85,215 @@ void whoami_nobody_okResponses() {
...
+                USERNAMES.get(1), USERNAMES.get(0));
         ObjectNode resp = runQuery("update-user", updateVariables);
         ObjectNode user = (ObjectNode) resp.get("updateUser");
         assertEquals("Ronda", user.get("firstName").asText());
         assertEquals(USERNAMES.get(1), user.get("email").asText());
         assertEquals(OrganizationRole.USER.getGrantedPermissions(), extractPermissionsFromUser(user));
     }
 
+    @Test
+    void updateUser_orgUser_failure() {
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:136]: ditto

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceEmptyImpl.java [hunk 22]
```diff
@@ -1,36 +1,54 @@
...
     public void createUser(IdentityAttributes userIdentity, String organizationExternalId) {}
 
     public void updateUser(String oldUsername, IdentityAttributes userIdentity) {}
 
-    public void createOrganization(String name, String externalId) {
+    public List<String> getAllUsernamesForOrganization(String externalId, OrganizationRole role) {
+        return orgRoleUsernamesMap.get(externalId).get(role);
     }
 
+    public void createOrganization(String name, String externalId) {}
```
[the remark this answers is not in the document]
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OktaServiceEmptyImpl.java]: yours and mine both!

**Comment by jeremyzitomer-usds** [comment]
@benwarfield-usds I've addressed your PR feedback. With the caveat that several of your suggestions are things I'm addressing in my next PR, please let me know if there are any outstanding concerns before I merge.

And--kindly ignore Sonar's complaints about code coverage. I think it's ignoring that the Okta live repository class was renamed, and doesn't like that that repository isn't unit tested (which, frankly, it should take up with Okta, not me). But including my tests in `ApiUserManagementTest.java`, I'm pretty darn sure relative code coverage for this PR is much higher than Sonar would have you think.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/OrganizationRole.java [hunk 23]
```diff
@@ -70,4 +69,8 @@ public int compare(OrganizationRole one, OrganizationRole other) {
             return Integer.compare(one.precedence, other.precedence);
         }
     }
+
+    public static final OrganizationRole getDefault() {
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/OrganizationRole.java:73]: 👍

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationInitializingService.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (benwarfield-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationInitializingService.java:89]: Ah, I see. I think that should probably not be the default for dev, but we can talk it through as we go.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/DemoOktaRepository.java [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+
+    public void updateUser(String oldUsername, IdentityAttributes userIdentity) {}
+
+    public List<String> getAllUsernamesForOrganization(Organization org, OrganizationRole role) {
+        return orgRoleUsernamesMap.getOrDefault(org.getExternalId(), new HashMap<>())
+                                  .getOrDefault(role, List.of());
+    }
+
+    public void createOrganization(String name, String externalId) {
+        orgRoleUsernamesMap.putIfAbsent(externalId, new HashMap<>());
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/DemoOktaRepository.java:77]: I... did not realize that method was in Map. 👍

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 25]
```diff
@@ -84,7 +100,22 @@ public void createUser(IdentityAttributes userIdentity, String organizationExter
                 .buildAndCreate(_client);
     }
 
+    public List<String> getAllUsernamesForOrganization(Organization org, OrganizationRole role) {
+        String groupName = generateGroupName(org.getExternalId(), role);
+
+        GroupList groups = _client.listGroups(groupName, null, null);
+        if (!groups.iterator().hasNext()) {
+            LOG.warn("Okta group is nonexistent; returning zero usernames");
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:108]: Maybe worth indicating which org and role the group is missing for?

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 26]
```diff
@@ -123,25 +154,37 @@ public void deleteOrganization(String externalId) {
...
+
         for (Group g : user.listGroups()) {
             String groupName = g.getProfile().getName();
-            // We assume that a user is only a member of one user group
-            if (g.getType() == GroupType.OKTA_GROUP &&
-                    groupName.startsWith(_rolePrefix) &&
-                    groupName.endsWith(generateRoleSuffix(OrganizationRole.USER))) {
-                return getOrganizationExternalIdFromGroupName(groupName, 
-                                                              OrganizationRole.USER);
+            for (OrganizationRole role : OrganizationRole.values()) {
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:170]: I think this code path and the one in OrganizationExtractor should not both exist in the long term, but I think this is in keeping with the other question about longer-term business rules for inactive orgs, and also potentially something we're going to refactor when we do V2 access control anyway.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/DemoOktaRepository.java [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+import gov.cdc.usds.simplereport.config.BeanProfiles;
+import gov.cdc.usds.simplereport.config.authorization.OrganizationRoleClaims;
+import gov.cdc.usds.simplereport.config.simplereport.DemoUserConfiguration;
+import gov.cdc.usds.simplereport.config.simplereport.DemoUserConfiguration.DemoUser;
+import gov.cdc.usds.simplereport.db.model.Organization;
+import gov.cdc.usds.simplereport.config.authorization.OrganizationRole;
+import gov.cdc.usds.simplereport.service.model.IdentityAttributes;
+
+/**
+ * Created by jeremyzitomer-usds on 1/7/21
```
Comment (benwarfield-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/DemoOktaRepository.java:25]: In the department of ridiculous nitpicks, I would turn off this auto-generation (or not type it in by hand)—it's all there in the git history anyway, and it's not really information that you need when using the public API of this class, which is what the javadoc should tell us.

File: backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java [hunk 28]
```diff
@@ -83,69 +89,191 @@ void whoami_nobody_okResponses() {
...
+        }
+        return list;
+    }
+
+    private OrganizationRoleClaims getOrgRoles(OrganizationRole role) {
+        Set<OrganizationRole> roles = new HashSet<>();
+        roles.add(OrganizationRole.USER);
+        roles.add(role);
+        return new OrganizationRoleClaims(_initService.getDefaultOrganization().getExternalId(),
+                                                   roles);
```
Comment (benwarfield-usds) [inline backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:242]:
```suggestion
        return new OrganizationRoleClaims(_initService.getDefaultOrganization().getExternalId(),
                                                   Set.of(OrganizationRole.USER, role));
```

Or not, honestly, this is fine. 🙂

File: backend/src/test/java/gov/cdc/usds/simplereport/idp/repository/OktaRepositoryTest.java [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+
+import org.springframework.beans.factory.annotation.Autowired;
+
+import gov.cdc.usds.simplereport.db.model.Organization;
+import gov.cdc.usds.simplereport.idp.repository.OktaRepository;
+import gov.cdc.usds.simplereport.config.authorization.OrganizationRole;
+import gov.cdc.usds.simplereport.config.authorization.OrganizationRoleClaims;
+import gov.cdc.usds.simplereport.service.model.IdentityAttributes;
+
+class OktaRepositoryTest {
```
Comment (benwarfield-usds) [inline backend/src/test/java/gov/cdc/usds/simplereport/idp/repository/OktaRepositoryTest.java:18]: I think it might be helpful to add a javadoc saying what these tests are and why we aren't running them.

I'd also love to get them running, at least optionally, but that's a separate question.

**Review by benwarfield-usds — APPROVED** [review]
LGTM!

I think.

That's a lot of code—congrats (possibly premature) on getting it over the finish line!

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java (hunk 26 above)
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:170]: Yeah, and this is designed specifically for users that are *not* the current user

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/DemoOktaRepository.java (hunk 27 above)
Comment (jeremyzitomer-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/DemoOktaRepository.java:25]: </3

File: backend/src/test/java/gov/cdc/usds/simplereport/idp/repository/OktaRepositoryTest.java (hunk 29 above)
Comment (jeremyzitomer-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/idp/repository/OktaRepositoryTest.java:18]: looking to get them running in v2

File: backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java (hunk 28 above)
Comment (jeremyzitomer-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java:242]: needs to be able to de-dupe duplicates though! Set can't, but HashSet can.

Comment (benwarfield-usds) [reply backend/src/test/java/gov/cdc/usds/simplereport/api/ApiUserManagementTest.java]: Really? Set.of chokes on duplicates? I would not have guessed that! (Obviously, I suppose!)
